webhook_handlers/views/gitlab.py

```diff
@@ -129,112 +155,92 @@ def _handle_merge_request_event(self, repo):
 
         return Response(data=message)
 
-    def _handle_system_hook_event(self, repo):
+    def _initiate_sync_for_owner(self, owner):
         """
-        GitLab Enterprise instance can send system hooks for changes on user, group, project, etc
-
-        http://doc.gitlab.com/ee/system_hooks/system_hooks.html
+        default: will sync_teams and sync_repos for owner
+        sync_teams to update owner.organizations list (expired memberships are removed and new memberships are added),
+        and username, name, email, and avatar of each Org in owner.organizations.
+        sync_repos to update owner.permission list (private repo access),
+        and name, language, private, repoid, and deleted=False for each repo the owner has access to.
         """
-        if not get_config("setup", "enterprise_license"):
-            raise PermissionDenied("No enterprise license detected")
-
-        event_name = self.request.data.get("event_name")
-        message = None
-
-        if event_name == "project_create":
-            owner_username, repo_name = self.request.data.get(
-                "path_with_namespace"
-            ).split("/", 2)
+        RefreshService().trigger_refresh(
+            ownerid=owner.ownerid,
+            username=owner.username,
+            using_integration=False,
+            manual_trigger=False,
+        )
 
+    def _try_initiate_sync_for_owner(self):
+        # could be the username of the OwnerUser or OwnerOrg. Sync only works with an OwnerUser.
+        owner_username_best_guess = self.request.data.get(
+            "path_with_namespace", ""
+        ).split("/")[0]
+        try:
+            owner = Owner.objects.get(
+                service=self.service_name,
+                oauth_token__isnull=False,
+                username=owner_username_best_guess,
+            )
+            self._initiate_sync_for_owner(owner)
+        except Owner.DoesNotExist:
+            owner_email = self.request.data.get("owner_email")
+            owner_name = self.request.data.get("owner_name")
             try:
                 owner = Owner.objects.get(
-                    service=self.service_name, username=owner_username
-                )
-
-                obj, created = Repository.objects.get_or_create(
-                    author=owner,
-                    service_id=self.request.data.get("project_id"),
-                    name=repo_name,
-                    private=self.request.data.get("project_visibility") == "private",
+                    service=self.service_name,
+                    oauth_token__isnull=False,
+                    name=owner_name,
+                    email=owner_email,
                 )
-                message = "Repository created"
+                self._initiate_sync_for_owner(owner)
```

Comment:
i think this could be one try/catch block?
```
try:
  owner = Owner.objects.get(
    service=self.service_name,
    oauth_token__isnull=False,
    username=owner_username_best_guess,
  )

  if not owner:
    owner_email = self.request.data.get("owner_email")
    owner_name = self.request.data.get("owner_name")
    owner = Owner.objects.get(
      service=self.service_name,
      oauth_token__isnull=False,
      name=owner_name,
      email=owner_email,
    )
except Owner.DoesNotExit:
    message = ":("
    return

self._initiate_sync_for_owner(owner)
```

should we just always use the `self.request` stuff or will it not always be there/correct?

Reply:
good call, I'll update 👍